### PR TITLE
Add more utilities to string arithmetic utility

### PR DESCRIPTION
**Pull request by ajreynol**
These will be required for a forthcoming MACRO theory rewrite that is required for scaling the RARE reconstruction for several classes of difficult string rewrites.

Relies on https://github.com/cvc5/cvc5/pull/10826, review that first.

### CODE REVIEW

File: src/theory/strings/arith_entail.cpp [hunk 1]
```diff
@@ -100,6 +101,116 @@ Node ArithEntail::rewriteArith(Node a)
...
+    it = visited.find(cur);
+    if (it == visited.end())
+    {
+      if (cur.getNumChildren() == 0)
+      {
+        visit.pop_back();
+        visited[cur] = cur;
+        continue;
+      }
+      visited[cur] = Node::null();
```
Comment (abdoo8080) [inline src/theory/strings/arith_entail.cpp:152]: Redundant? Use emplace

File: src/theory/strings/arith_entail.cpp [hunk 2]
```diff
@@ -100,6 +101,116 @@ Node ArithEntail::rewriteArith(Node a)
...
+Node ArithEntail::rewriteLengthIntro(const Node& n,
+                                     TConvProofGenerator* pg) const
+{
+  NodeManager* nm = NodeManager::currentNM();
+  std::unordered_map<TNode, Node> visited;
+  std::unordered_map<TNode, Node>::iterator it;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
```
Comment (abdoo8080) [inline src/theory/strings/arith_entail.cpp:140]: Is this loop a performance bottleneck? It's quite complicated for what it is doing... I would just split it into two loops: the first one for getting a stack of children and the second one for processing them.

Comment (ajreynol) [reply src/theory/strings/arith_entail.cpp:140]: This method is potentially a bottleneck.

The problem is that "processing" them means replacing the subterm at that point. One would actually need 2 traversals and 1 loop to process (a traversal to collect, process, another traversal to construct the term). So I'd prefer keeping a single traversal for performance.

File: src/theory/strings/arith_entail.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/strings/arith_entail.cpp:152]: visited.empace(cur)?

**Review by abdoo8080 — APPROVED** [review]

File: src/theory/strings/arith_entail.cpp (hunk 1 above)
Comment (abdoo8080) [reply src/theory/strings/arith_entail.cpp:152]: `visited.emplace(cur, Node::null());`

**Pull request merged**